test/optimizer/optimizer_rule_test.cpp

```diff
@@ -62,6 +65,199 @@ TEST_F(OptimizerRuleTests, SimpleRuleApplyTest) {
+  child_join->PushChild(left_leaf);
+  child_join->PushChild(middle_leaf);
+  optimizer.metadata_.memo.InsertExpression(
+      optimizer.metadata_.MakeGroupExpression(child_join), true);
+
+  // Make Parent join
+  std::vector<AnnotatedExpression> parent_join_predicates;
+  std::unordered_set<std::string> parent_tables({"test1", "test3"});
+  pred = {dummy_expr, parent_tables};
+  parent_join_predicates.push_back(pred);
+
+  auto parent_join = std::make_shared<OperatorExpression>(
+      LogicalInnerJoin::make(parent_join_predicates));
+  parent_join->PushChild(child_join);
+  parent_join->PushChild(right_leaf);
+
+  optimizer.metadata_.memo.InsertExpression(
+      optimizer.metadata_.MakeGroupExpression(parent_join), true);
+  OptimizeContext *root_context =
+      new OptimizeContext(&(optimizer.metadata_), nullptr);
+
+  EXPECT_EQ(left_leaf, parent_join->Children()[0]->Children()[0]);
+  EXPECT_EQ(middle_leaf, parent_join->Children()[0]->Children()[1]);
+  EXPECT_EQ(right_leaf, parent_join->Children()[1]);
+  EXPECT_EQ(1,
+            parent_join->Op().As<LogicalInnerJoin>()->join_predicates.size());
+  EXPECT_EQ(1, parent_join->Children()[0]
+                   ->Op()
+                   .As<LogicalInnerJoin>()
+                   ->join_predicates.size());
+
+  // Setup rule
+  InnerJoinAssociativity rule;
+
+  EXPECT_TRUE(rule.Check(parent_join, root_context));
+  std::vector<std::shared_ptr<OperatorExpression>> outputs;
+  rule.Transform(parent_join, outputs, root_context);
+  EXPECT_EQ(1, outputs.size());
+
+  auto output_join = outputs[0];
+
+  EXPECT_EQ(left_leaf, output_join->Children()[0]);
+  EXPECT_EQ(middle_leaf, output_join->Children()[1]->Children()[0]);
+  EXPECT_EQ(right_leaf, output_join->Children()[1]->Children()[1]);
+
+  auto parent_join_op = output_join->Op().As<LogicalInnerJoin>();
+  auto child_join_op = output_join->Children()[1]->Op().As<LogicalInnerJoin>();
+  EXPECT_EQ(2, parent_join_op->join_predicates.size());
+  EXPECT_EQ(0, child_join_op->join_predicates.size());
+}
+
+TEST_F(OptimizerRuleTests, SimpleAssociativeRuleTest2) {
+  // Start Join Structure: (left JOIN middle) JOIN right
+  // End Join Structure: left JOIN (middle JOIN right)
+  // Query: SELECT * from test1, test2, test3 WHERE test1.a = test3.a AND
+  // test2.a = test3.a; Test ensures that predicate "test2.a = test3.a" is
+  // redistributed to child join
+
+  // Setup Memo
+  Optimizer optimizer;
+  auto &memo = optimizer.metadata_.memo;
+
+  auto left_get = std::make_shared<OperatorExpression>(
+      LogicalGet::make(0, {}, nullptr, "test1"));
+  auto middle_get = std::make_shared<OperatorExpression>(
+      LogicalGet::make(1, {}, nullptr, "test2"));
+  auto right_get = std::make_shared<OperatorExpression>(
+      LogicalGet::make(2, {}, nullptr, "test3"));
+
+  // Create Groups for Get Operators
+  auto left_get_group = memo.InsertExpression(
+      optimizer.metadata_.MakeGroupExpression(left_get), true);
```

Comment:
I actually don't know whether it should be enforced/not enforced... I just want to make sure. @chenboy Can you explain this one a bit as well?

Reply:
It should not. We only enforce physical properties, e.g. we treat`PhysicalOrderBy` as enforced operator when adding sort property. `LogicalGet` should not be created as enforced here.

Reply:
Do we ever use the enforced/not enforced information in the later stage of the optimization pipeline?

Reply:
Yes, we will not enumerate enforced expressions when finding the best expression in a group. This is because we will enforce properties after this step anyway. Another reason is that enforced expressions have its own group as its child. When costing the expression, we may need to cost its child group first, this may cause an endless loop since it would cost itself again and again. We think separating enforced expressions is the clearest way.

Reply:
I see. Thanks for the clarification!